### DISCUSSION TITLE
[DSL] Types and Typechecking

### DISCUSSION THREAD

**Discussion by malt-r** [Programmiermethoden/Dungeon#1375] [Ideas]
Status Quo:
Die DSL hat ein statisches Typsystem. Die Motivation hierfür war, bereits vor der Ausführung der DSL-Programme möglichst aussagekräftige Fehlermeldungen generieren zu können. 
Das ist aus Endnutzendensicht wichtig, um möglichst früh auf Fehlerfälle hingewiesen zu werden. Aus Entwickelndensicht ist es allerdings auch wichtig, da sich aktuell die Fehler, die auf Typ-Mismatches zurückzuführen sind, erst in der Laufzeit manifestieren, was zu unnötig langen Fehlersuchen führt. Es wird aktuell in der semantischen Analyse kein Typechecking und keine Typinferenz umgesetzt. 

Es muss festgelegt werden, welche Regeln für die Typinferenzberechnung gelten sollen und wann Symbole unterschiedlicher Datentypen einander zugewiesen werden können (z. B. `int`-Wert einer `float`-Variable zuweisen). Für das Typechecking muss zusammen mit dem Themenkomplex [Errorhandling](https://github.com/Programmiermethoden/Dungeon/issues/56) festgelegt werden, wie Typfehler behandelt und kommuniziert werden sollen.

**Comment by cagix** [comment]
wird irgendwie in der geplanten thesis von @malt-r auch mit berührt. falls nicht -> schließen als "not planned".
